holoviews/element/raster.py

```diff
@@ -408,45 +211,157 @@ class Image(SheetCoordinateSystem, Raster):
 
-
     def __init__(self, data, bounds=None, extents=None, xdensity=None, ydensity=None, **params):
-        bounds = bounds if bounds is not None else BoundingBox()
-        if np.isscalar(bounds):
+        extents = extents if extents else (None, None, None, None)
+        if data is None: data = np.array([[0]])
+        Dataset.__init__(self, data, extents=extents, bounds=None, **params)
+
+        if isinstance(data, np.ndarray):
+            dim2, dim1 = data.shape[:2]
+        else:
+            dim1, dim2 = tuple([len(self.dimension_values(kd, expanded=False))
+                                  for kd in self.kdims])
+        if self.bounds is not None:
+            bounds = self.bounds
+        elif bounds is None and isinstance(data, np.ndarray):
+            bounds = BoundingBox()
+
+        if bounds is None:
+            xvals = self.dimension_values(0, False)
+            l, r, xdensity, _ = util.bound_range(xvals, xdensity)
+            yvals = self.dimension_values(1, False)
+            b, t, ydensity, _ = util.bound_range(yvals, ydensity)
+            bounds = BoundingBox(points=((l, b), (r, t)))
+        elif np.isscalar(bounds):
             bounds = BoundingBox(radius=bounds)
         elif isinstance(bounds, (tuple, list, np.ndarray)):
             l, b, r, t = bounds
             bounds = BoundingBox(points=((l, b), (r, t)))
-        if data is None: data = np.array([[0]])
-        l, b, r, t = bounds.lbrt()
-        extents = extents if extents else (None, None, None, None)
-        Element2D.__init__(self, data, extents=extents, bounds=bounds,
-                           **params)
 
-        (dim1, dim2) = self.data.shape[1], self.data.shape[0]
+        l, b, r, t = bounds.lbrt()
         xdensity = xdensity if xdensity else dim1/float(r-l)
         ydensity = ydensity if ydensity else dim2/float(t-b)
         SheetCoordinateSystem.__init__(self, bounds, xdensity, ydensity)
 
-        if len(self.data.shape) == 3:
-            if self.data.shape[2] != len(self.vdims):
+        if len(self.shape) == 3:
+            if self.shape[2] != len(self.vdims):
                 raise ValueError("Input array has shape %r but %d value dimensions defined"
-                                 % (self.data.shape, len(self.vdims)))
+                                 % (self.shape, len(self.vdims)))
+
+
+    def __setstate__(self, state):
+        """
+        Ensures old-style unpickled Image types without an interface
+        use the ImageInterface.
+        """
+        self.__dict__ = state
+        if isinstance(self.data, np.ndarray):
+            self.interface = ImageInterface
+        super(Dataset, self).__setstate__(state)
+
+
+    def aggregate(self, dimensions=None, function=None, spreadfn=None, **kwargs):
+        agg = super(Image, self).aggregate(dimensions, function, spreadfn, **kwargs)
+        return Curve(agg) if isinstance(agg, Dataset) and len(self.vdims) == 1 else agg
```

Comment:
I thought aggregate would already pick the appropriate element - or is that part of the proposed reductions PR?